server/src/crypto.rs

```diff
@@ -0,0 +1,29 @@
+use crate::{
+    platform::{
+        crypto::{Drbg, EntropySource},
+        TpmContextDeps,
+    },
+    ServerError,
+};
+
+pub struct Crypto<Deps: TpmContextDeps> {
+    pub drbg: Deps::Drbg,
+    pub entropy: Deps::EntropySource,
+}
+
+impl<Deps: TpmContextDeps> Crypto<Deps> {
```

Comment:
I'm still unclear on what's going on here.

My original expectation was that there'd only be one DRBG, is that the intent? If so it's a bit weird that there's this hierarchy of `Crypto` > `CommandHandler` > `TpmContext`.

`TpmContext` suggests a single instance, but `CommandHandler` (by name) suggests an instance per command. So I'm left wondering what the intent here is. I think this gets into my original issue I reported; https://github.com/tpm-rs/tpm-rs/issues/141 - so maybe this is a duplicate concern here.

---

Anyways, let's say that my original interpretation of intent was correct, and that we only want one `Drbg` per instance of `TpmContext`. In that case, all these `new` functions, and all the traits with the ability to `initialize` feel unneeded to me, because I would have much rather the manufacturer constructed the object and passed it to us, and that we don't know how to construct it.

Like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fc242e13de088650f45b11cb1c89b73b

Reply:
No I think your interpretation does not align with what i had in mind.

I was intending for TpmContext to be the top user facing TPM implementation, probably managing IO, CommandHandler is the lower below which handles all commands in the single instance, I have couple of sketches based on your ideas but I will send them later, Now Crypto is the platform API for cryptography including DRBG and later will include full cryptography suite  (AES, RSA, Hashes, ECC, ...etc).

It may not be the best, but it can be always changed later, I didn't put time to document that because it is just organically developed at this point,   and I am mainly focusing on the DRBG interface aspect.

Reply:
Okay, so - `CommandHandler` is more like some kind of thing that just bundles resources, and there's not one of them per command.

In that case it *is* what I hoped for in my original comment here. In that case, we don't need the TPM Server code to know how to build these, and we can instead ask the person constructing the TPM server instance to construct it - so we don't really *need* any of these `new` and/or `initialize` functions in the future.

It's fine if we want to do this for now; I'll approve with the expectation that more design will be put into this in the future. I strongly suspect we don't need `Drbg::initialize` or `Drbg::Error` or `Drbg::Entropy` or `Drbg::Nonce`, as I've commented elsewhere - but maybe it's a tall order to ask they be removed this PR.

Reply:
Thanks alot for your support, I will fix some of the issues you raised, and merge tomorrow.

> so we don't really need any of these new and/or initialize functions in the future.

I had a dicussion with a collague about this, and from what I understand this is actually necessary, but it is not obvious in this case of generate random. The problem is for deriving keys, the tpm would use a different instance (possibly short lived) of the DRBG, only difference is this time it will be seeded by static seed.

It should be easy to remove `Drbg::Error` and I will do that. As for `Entropy` and `Nonce` They are a bit tricky to remove (in particular regarding the real DRBG implementation) but we can revisit later.

I will fix the minor issues and merge tomorrow. Once again thanks for your support!